examples/bone_finetuning/finetune.py

```diff
@@ -0,0 +1,385 @@
+# Copyright 2024-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import logging
+import os
+import random
+from dataclasses import dataclass, field
+from typing import Dict, List, Literal, Optional, Sequence
+
+import datasets
+import numpy as np
+import torch
+import torch.distributed
+import transformers
+from datasets import load_dataset
+from transformers import BitsAndBytesConfig, Trainer
+from transformers.trainer_utils import PREFIX_CHECKPOINT_DIR
+
+from peft import BoneConfig, LoraConfig, PeftModel, TaskType, get_peft_model, prepare_model_for_kbit_training
+
+
+IGNORE_INDEX = -100
+EOT_TOKEN = "<|EOT|>"
+logger = logging.getLogger(__name__)
+
+PROMPT = (
+    "Below is an instruction that describes a task. "
+    "Write a response that appropriately completes the request.\n\n"
+    "### Instruction:\n{instruction}\n\n### Response:"
+)
+
+
+@dataclass
+class TrainingArguments(transformers.TrainingArguments):
+    # Base model or residual model setting
+    model_name_or_path: Optional[str] = field(default="meta-llama/Meta-Llama-3-8B")
+    attn_implementation: Optional[str] = field(default="flash_attention_2")
+    # Bone setting
+    use_bone: Optional[bool] = field(default=False)
+    bone_r: Optional[int] = field(default=64)
+    init_bone_weights: Literal[True, False] = field(default=True)
+    # Lora or PiSSA setting
+    use_lora: Optional[bool] = field(default=False)
+    adapter_name_or_path: Optional[str] = field(
+        default=None,
+        metadata={
+            "help": (
+                "Pre-initialized Bone adapter path; when this is not None, the following arguments are ignored."
+            ),
+        },
+    )
+    init_lora_weights: Literal[True, "pissa_niter_4", "olora"] = field(
+        default=True,
+        metadata={
+            "help": ("True -> LoRA; `pissa` -> PiSSA"),
+        },
+    )
+    target_modules: Optional[str] = field(default="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj")
+    lora_rank: Optional[int] = field(default=8)
+    lora_alpha: Optional[float] = field(default=32.0)
+    lora_dropout: Optional[float] = field(
+        default=0.0,
+        metadata={
+            "help": ("Must be set to 0 when using PiSSA."),
+        },
+    )
+    # Quantization setting
+    bits: int = field(default=16, metadata={"help": "How many bits to use."})
```

Comment:
I tried running the script with Bone and `bits=8` and this raised an error. This is probably not surprising, since Bone does not explicitly support bitsandbytes, right? So how about adding a check that if Bone is being used, `bits` should be 16 or 32.